src/engine/input/key-mapping/InputEventItem.cs

```diff
@@ -0,0 +1,138 @@
+﻿using System;
+using Godot;
+
+public class InputEventItem : Node
+{
+    [Export]
+    public NodePath ButtonPath;
+
+    internal InputActionItem AssociatedAction;
+
+    internal bool JustAdded;
+
+    private Button button;
+
+    private bool waitingForInput;
+
+    public InputEventWithModifiers AssociatedEvent { get; set; }
+
+    public void Delete()
+    {
+        AssociatedAction.AssociatedGroup.GetParent<InputGroupList>().ControlSchemeChanged();
+        AssociatedAction.Inputs.Remove(this);
+    }
+
+    public override void _Input(InputEvent @event)
+    {
+        if (InputGroupList.Instance.IsConflictDialogOpen())
+            return;
+
+        if (!waitingForInput)
+            return;
+
+        if (!(@event is InputEventMouseButton) && !(@event is InputEventKey))
+            return;
+
+        if (@event is InputEventKey iek)
+        {
+            switch (iek.Scancode)
+            {
+                case (uint)KeyList.Escape:
+                    Delete();
+                    return;
+                case (uint)KeyList.Alt:
+                case (uint)KeyList.Shift:
+                case (uint)KeyList.Control:
+                    return;
+            }
+        }
+
+        var old = AssociatedEvent;
+        AssociatedEvent = (InputEventWithModifiers)@event;
+        var conflict = InputGroupList.Instance.Conflicts(this);
+        if (conflict != null)
+        {
+            InputGroupList.Instance.ShowDialog(this, conflict, AssociatedEvent);
+            AssociatedEvent = old;
+            return;
+        }
+
+        waitingForInput = false;
+        JustAdded = false;
+        AssociatedAction.AssociatedGroup.GetParent<InputGroupList>().ControlSchemeChanged();
+
+        UpdateButtonText();
+    }
+
+    public override void _Ready()
+    {
+        button = GetNode<Button>(ButtonPath);
+        if (JustAdded)
+            OnButtonPressed();
+        else
+            UpdateButtonText();
+    }
+
+    public override bool Equals(object obj)
```

Comment:
Are all of these custom Equals and GetHashCode methods needed?

Reply:
they are required by the Conflicts method